### PR TITLE
bugfix #509: fixed rating pagination

### LINKED ISSUE #509: Fix rating pagination
There is a bug associated with slow loading of teachers and grades on teacher and subject pages. After all, with the increase in the number of teachers, the number of requests increases.

So you need to fix the pagination of the rating to fix this.

### PR DESCRIPTION
**Pull request by f-ogoog**
- now one large request is sent to the database instead of many small ones

closes #509

### CODE REVIEW

File: src/v2/api/services/SubjectService.ts [hunk 1]
```diff
@@ -80,17 +83,41 @@ export class SubjectService {
...
-    };
 
     for (const dbTeacher of dbTeachers) {
+      const sortedQuestionsWithAnswers = this.questionMapper.getSortedQuestionsWithAnswers(dbTeacher.disciplineTeachers);
+      const marks = sortedQuestionsWithAnswers.map((q) => ({
+        name: q.name,
+        amount: q.questionAnswers.length,
+        type: q.display,
+        mark: this.teacherService.getRightMarkFormat(q),
+      }));
```
Comment (basarabst) [inline src/v2/api/services/SubjectService.ts:116]: move it to mapper

File: src/v2/api/services/TeacherService.ts [hunk 2]
```diff
@@ -56,25 +59,45 @@ export class TeacherService {
...
+    const teachers = await this.getAll(body);
+    const teachersWithRating = [];
+    for (const { disciplineTeachers, ...dbTeacher } of teachers.data) {
+      const sortedQuestionsWithAnswers = this.questionMapper.getSortedQuestionsWithAnswers(disciplineTeachers);
+      const marks = sortedQuestionsWithAnswers.map((q) => ({
+        name: q.name,
+        amount: q.questionAnswers.length,
+        type: q.display,
+        mark: this.getRightMarkFormat(q),
+      }));
```
Comment (basarabst) [inline src/v2/api/services/TeacherService.ts:90]: move it to mapper to avoid duplication

**Review by basarabst — CHANGES_REQUESTED** [review]

**Review by Hahhhayam — APPROVED** [review]

**Review by basarabst — APPROVED** [review]

File: src/v2/api/services/SubjectService.ts [hunk 3]
```diff
@@ -80,17 +83,36 @@ export class SubjectService {
           },
         },
       },
+      include: {
```
Comment (just-hoshinon) [inline src/v2/api/services/SubjectService.ts:86]: Where is the new type of this sctructure?

File: src/v2/mappers/QuestionMapper.ts [hunk 4]
```diff
@@ -87,4 +87,54 @@ export class QuestionMapper {
     }
     return responses;
   }
+
+  getSortedQuestionsWithAnswers (disciplineTeachers) {
```
Comment (just-hoshinon) [inline src/v2/mappers/QuestionMapper.ts:91]: where is the type

File: src/v2/mappers/QuestionMapper.ts [hunk 5]
```diff
@@ -87,4 +87,54 @@ export class QuestionMapper {
...
+    const sortedQuestions = [];
+    for (const disciplineTeacher of disciplineTeachers) {
+      for (const questionWithAnswer of disciplineTeacher.questionAnswers) {
+        const { question, value } = questionWithAnswer;
+        if (question.type === QuestionType.TEXT) continue;
```
Comment (just-hoshinon) [inline src/v2/mappers/QuestionMapper.ts:96]: you can do filtration in the database request

File: src/v2/mappers/QuestionMapper.ts [hunk 6]
```diff
@@ -87,4 +87,54 @@ export class QuestionMapper {
...
+      mark: this.getRightMarkFormat(q),
+    }));
+  }
+
+  private parseMark (type: QuestionType, marksSum: number, answerQty: number) {
+    const divider = (answerQty * ((type === QuestionType.SCALE) ? 10 : 1));
+    return parseFloat(((marksSum / divider) * 100).toFixed(2));
+  }
+
+  getRightMarkFormat ({ display, type, questionAnswers: answers }) {
```
Comment (just-hoshinon) [inline src/v2/mappers/QuestionMapper.ts:129]: where is the type

File: src/v2/api/services/TeacherService.ts [hunk 7]
```diff
@@ -56,25 +59,40 @@ export class TeacherService {
...
+            roles: true,
+            questionAnswers: {
+              include: {
+                question: true,
+              },
+            },
+          },
+        },
+      },
     };
```
Comment (just-hoshinon) [inline src/v2/api/services/TeacherService.ts:75]: Where is the type

File: src/v2/mappers/QuestionMapper.ts [hunk 8]
```diff
@@ -87,4 +87,54 @@ export class QuestionMapper {
...
+          });
+        } else {
+          sortedQuestion.questionAnswers.push({ value });
+        }
+      }
+    }
+    return sortedQuestions;
+  }
+
+  getMarks (questionWithAnswers) {
```
Comment (just-hoshinon) [inline src/v2/mappers/QuestionMapper.ts:115]: Where is the type

**Review by just-hoshinon — CHANGES_REQUESTED** [review]

**Review by just-hoshinon — APPROVED** [review]

**Pull request merged**